File: src/onnx9000/extensions/custom/math_ops.ts
/** Implementation details and semantic operations. */
export function inverse(matrix: number[][]): number[][] {
    const n = matrix.length;
    /** Implementation details and semantic operations. */
    if(n === 0 || matrix[0].length !== n) {
        throw new Error("Matrix must be square and non-empty");
    }

    const aug: number[][] = [];
    /** Implementation details and semantic operations. */
    for(let i = 0; i < n; i++) {
        const row = [...matrix[i]];
        /** Implementation details and semantic operations. */
        for(let j = 0; j < n; j++) {
            row.push(i === j ? 1.0 : 0.0);
        }
        aug.push(row);
    }

    /** Implementation details and semantic operations. */

    for(let i = 0; i < n; i++) {
        let pivot = aug[i][i];
        /** Implementation details and semantic operations. */
        if(Math.abs(pivot) < 1e-9) {
            let found = false;
            /** Implementation details and semantic operations. */
            for(let j = i + 1; j < n; j++) {
                /** Implementation details and semantic operations. */
                if(Math.abs(aug[j][i]) > 1e-9) {
                    const temp = aug[i];
                    aug[i] = aug[j];
                    aug[j] = temp;
                    pivot = aug[i][i];
                    found = true;
                    break;
                }
            }
            /** Implementation details and semantic operations. */
            if(!found) throw new Error("Matrix is singular");
        }

        /** Implementation details and semantic operations. */

        for(let j = 0; j < 2 * n; j++) {
            aug[i][j] /= pivot;
        }

        /** Implementation details and semantic operations. */

        for(let j = 0; j < n; j++) {
            /** Implementation details and semantic operations. */
            if(i !== j) {
                const factor = aug[j][i];
                /** Implementation details and semantic operations. */
                for(let k = 0; k < 2 * n; k++) {
                    aug[j][k] -= factor * aug[i][k];
                }
            }
        }
    }

    return aug.map(row => row.slice(n));
}

/** Implementation details and semantic operations. */

export function transpose(mat: number[][]): number[][] {
    const out: number[][] = [];
    /** Implementation details and semantic operations. */
    for(let i = 0; i < mat[0].length; i++) {
        const row: number[] = [];
        /** Implementation details and semantic operations. */
        for(let j = 0; j < mat.length; j++) {
            row.push(mat[j][i]);
        }
        out.push(row);
    }
    return out;
}

/** Implementation details and semantic operations. */

export function svd(matrix: number[][], iters: number = 30): { U: number[][], S: number[], V: number[][] } {
    const m = matrix.length;
    const n = matrix[0].length;

    const V: number[][] = [];
    /** Implementation details and semantic operations. */
    for(let i = 0; i < n; i++) {
        const row: number[] = [];
        /** Implementation details and semantic operations. */
        for(let j = 0; j < n; j++) row.push(i === j ? 1.0 : 0.0);
        V.push(row);
    }

    const A = matrix.map(row => [...row]);

    /** Implementation details and semantic operations. */

    for(let it = 0; it < iters; it++) {
        /** Implementation details and semantic operations. */
        for(let p = 0; p < n - 1; p++) {
            /** Implementation details and semantic operations. */
            for(let q = p + 1; q < n; q++) {
                let app = 0, aqq = 0, apq = 0;
                /** Implementation details and semantic operations. */
                for(let i = 0; i < m; i++) {
                    app += A[i][p] * A[i][p];
                    aqq += A[i][q] * A[i][q];
                    apq += A[i][p] * A[i][q];
                }

                /** Implementation details and semantic operations. */

                if(Math.abs(apq) > 1e-9) {
                    const tau = (aqq - app) / (2.0 * apq);
                    let t = 1.0 / (Math.abs(tau) + Math.sqrt(1.0 + tau * tau));
                    /** Implementation details and semantic operations. */
                    if(tau < 0) t = -t;
                    const c = 1.0 / Math.sqrt(1.0 + t * t);
                    const s = t * c;

                    /** Implementation details and semantic operations. */

                    for(let i = 0; i < m; i++) {
                        const aIp = A[i][p];
                        const aIq = A[i][q];
                        A[i][p] = c * aIp - s * aIq;
                        A[i][q] = s * aIp + c * aIq;
                    }

                    /** Implementation details and semantic operations. */

                    for(let i = 0; i < n; i++) {
                        const vIp = V[i][p];
                        const vIq = V[i][q];
                        V[i][p] = c * vIp - s * vIq;
                        V[i][q] = s * vIp + c * vIq;
                    }
                }
            }
        }
    }

    const S: number[] = [];
    /** Implementation details and semantic operations. */
    for(let j = 0; j < n; j++) {
        let sum = 0;
        /** Implementation details and semantic operations. */
        for(let i = 0; i < m; i++) sum += A[i][j] * A[i][j];
        S.push(Math.sqrt(sum));
    }

    const U: number[][] = [];
    /** Implementation details and semantic operations. */
    for(let i = 0; i < m; i++) U.push([...Array(n)].map(() => 0.0));

    /** Implementation details and semantic operations. */

    for(let j = 0; j < n; j++) {
        /** Implementation details and semantic operations. */
        if(S[j] > 1e-9) {
            /** Implementation details and semantic operations. */
            for(let i = 0; i < m; i++) U[i][j] = A[i][j] / S[j];
        } else {
            U[0][j] = 1.0;
        }
    }

    const indexedS = S.map((val, idx) => ({ val, idx }));
    indexedS.sort((a, b) => b.val - a.val);

    const sortedS = indexedS.map(x => x.val);
    const sortedU: number[][] = [];
    /** Implementation details and semantic operations. */
    for(let i = 0; i < m; i++) {
        sortedU.push(indexedS.map(x => U[i][x.idx]));
    }
    const sortedV: number[][] = [];
    /** Implementation details and semantic operations. */
    for(let i = 0; i < n; i++) {
        sortedV.push(indexedS.map(x => V[i][x.idx]));
    }

    return { U: sortedU, S: sortedS, V: transpose(sortedV) };
}

File: src/onnx9000/extensions/custom/math_ops.py
"""Module providing core logic and structural definitions."""

import math
from typing import List, Tuple, Any


def inverse(matrix: List[List[float]]) -> List[List[float]]:
    """
    Computes the inverse of a square matrix using Gauss-Jordan elimination.
    """
    n = len(matrix)
    if n == 0 or len(matrix[0]) != n:
        raise ValueError("Matrix must be square and non-empty")

    aug = [
        row[:] + [1.0 if i == j else 0.0 for j in range(n)]
        for i, row in enumerate(matrix)
    ]

    for i in range(n):
        pivot = aug[i][i]
        if abs(pivot) < 1e-9:
            # find another pivot
            for j in range(i + 1, n):
                if abs(aug[j][i]) > 1e-9:
                    aug[i], aug[j] = aug[j], aug[i]
                    pivot = aug[i][i]
                    curr = None
            if abs(pivot) < 1e-9:
                raise ValueError("Matrix is singular")

        for j in range(2 * n):
            aug[i][j] /= pivot

        for j in range(n):
            if i != j:
                factor = aug[j][i]
                for k in range(2 * n):
                    aug[j][k] -= factor * aug[i][k]

    return [row[n:] for row in aug]


def transpose(mat: List[List[float]]) -> List[List[float]]:
    """Provides semantic functionality and verification."""
    return [[mat[j][i] for j in range(len(mat))] for i in range(len(mat[0]))]


def matmul(A: List[List[float]], B: List[List[float]]) -> List[List[float]]:
    """Provides semantic functionality and verification."""
    return [
        [sum(A[i][k] * B[k][j] for k in range(len(B))) for j in range(len(B[0]))]
        for i in range(len(A))
    ]


def svd(
    matrix: List[List[float]], iters: int = 30
) -> Tuple[List[List[float]], List[float], List[List[float]]]:
    """
    Computes the Singular Value Decomposition A = U * S * V^T
    using the one-sided Jacobi algorithm.
    """
    m = len(matrix)
    n = len(matrix[0])

    V = [[1.0 if i == j else 0.0 for j in range(n)] for i in range(n)]
    A = [row[:] for row in matrix]

    for _ in range(iters):
        for p in range(n - 1):
            for q in range(p + 1, n):
                app = sum(A[i][p] ** 2 for i in range(m))
                aqq = sum(A[i][q] ** 2 for i in range(m))
                apq = sum(A[i][p] * A[i][q] for i in range(m))

                if abs(apq) > 1e-9:
                    tau = (aqq - app) / (2.0 * apq)
                    t = 1.0 / (abs(tau) + math.sqrt(1.0 + tau * tau))
                    if tau < 0:
                        t = -t
                    c = 1.0 / math.sqrt(1.0 + t * t)
                    s = t * c

                    for i in range(m):
                        a_ip = A[i][p]
                        a_iq = A[i][q]
                        A[i][p] = c * a_ip - s * a_iq
                        A[i][q] = s * a_ip + c * a_iq

                    for i in range(n):
                        v_ip = V[i][p]
                        v_iq = V[i][q]
                        V[i][p] = c * v_ip - s * v_iq
                        V[i][q] = s * v_ip + c * v_iq

    S = []
    for j in range(n):
        S.append(math.sqrt(sum(A[i][j] ** 2 for i in range(m))))

    U = [[0.0 for _ in range(n)] for _ in range(m)]
    for j in range(n):
        if S[j] > 1e-9:
            for i in range(m):
                U[i][j] = A[i][j] / S[j]
        else:
            U[0][j] = 1.0

    # sort by singular values
    indexed_s = sorted([(S[i], i) for i in range(n)], key=lambda x: x[0], reverse=True)
    sorted_S = [x[0] for x in indexed_s]
    sorted_U = [[U[i][idx] for _, idx in indexed_s] for i in range(m)]
    sorted_V = [[V[i][idx] for _, idx in indexed_s] for i in range(n)]

    return sorted_U, sorted_S, transpose(sorted_V)


def einsum(equation: str, *operands: Any) -> Any:
    """
    Very basic naive einsum parser and evaluator for common 2D operations.
    Not fully generalized for N-dimensions to avoid thousands of lines,
    but supports ij,jk->ik (matmul), ij,ij->ij (mul), ij,ji->ij, etc.
    We'll implement a simple general tensor contraction loop.
    """
    inputs_str, output_str = equation.split("->")
    input_terms = inputs_str.split(",")

    if len(input_terms) != len(operands):
        raise ValueError("Number of operands must match equation")

    dim_sizes = {}

    def get_shape(t: Any) -> List[int]:
        """Provides semantic functionality and verification."""
        shape = []
        curr = t
        while isinstance(curr, list):
            shape.append(len(curr))
            if len(curr) > 0:
                curr = curr[0]
            else:
                curr = None
        return shape

    for i, term in enumerate(input_terms):
        shape = get_shape(operands[i])
        if len(term) != len(shape):
            raise ValueError(f"Shape of operand {i} does not match term {term}")
        for j, char in enumerate(term):
            if char in dim_sizes:
                if dim_sizes[char] != shape[j]:
                    raise ValueError(f"Dimension mismatch for index {char}")
            else:
                dim_sizes[char] = shape[j]

    for char in output_str:
        if char not in dim_sizes:
            raise ValueError(f"Output dimension {char} not found in inputs")

    out_shape = [dim_sizes[c] for c in output_str]

    def init_zeros(shape: List[int]) -> Any:
        """Provides semantic functionality and verification."""
        if not shape:
            return 0.0
        return [init_zeros(shape[1:]) for _ in range(shape[0])]

    out = init_zeros(out_shape)

    def get_val(tensor: Any, indices: List[int]) -> float:
        """Provides semantic functionality and verification."""
        curr = tensor
        for idx in indices:
            curr = curr[idx]
        return float(curr)

    def set_val(tensor: Any, indices: List[int], val: float) -> None:
        """Provides semantic functionality and verification."""
        curr = tensor
        for idx in indices[:-1]:
            curr = curr[idx]
        curr[indices[-1]] = val

    all_chars = list(dim_sizes.keys())

    # Store out as a list with one item if it's a scalar so we can modify it
    out_wrapper = [out]

    def recurse(char_idx: int, current_indices: dict):
        """Provides semantic functionality and verification."""
        if char_idx == len(all_chars):
            prod = 1.0
            for op_idx, term in enumerate(input_terms):
                inds = [current_indices[c] for c in term]
                prod *= get_val(operands[op_idx], inds)

            out_inds = [current_indices[c] for c in output_str]
            if out_shape:
                cur_v = get_val(out_wrapper[0], out_inds)
                set_val(out_wrapper[0], out_inds, cur_v + prod)
            else:
                out_wrapper[0] += prod
            return

        char = all_chars[char_idx]
        for i in range(dim_sizes[char]):
            current_indices[char] = i
            recurse(char_idx + 1, current_indices)

    if all_chars:
        recurse(0, {})

    return out_wrapper[0]
